### PR TITLE
Feature/update data owner

### PR DESCRIPTION
**Pull request by ivanlele**

### CODE REVIEW

File: operation-update-data-owner.x [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    NOT_FOUND = -2,
+    //: Only owner can update data entry
+    NOT_AUTHORIZED = -3
+};
+
+//: Result of operation application
+union UpdateDataOwnerResult switch (UpdateDataOwnerResultCode code)
+{
+case SUCCESS:
+    //: Reserved for future extension
```
Comment (1henny) [inline operation-update-data-owner.x]: here we should provide new owner's ID

File: operation-update-data-owner.x [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace stellar
+{
+
+struct UpdateDataOwnerOp
+{
+    //: ID of the data entry to update an owner
+    uint64 dataID;
+    //: A new owner of the entry
+    AccountID owner;
```
Comment (1henny) [inline operation-update-data-owner.x]: let's name it `destination` or `new_owner` as it is more descriptive

Comment (1henny) [reply operation-update-data-owner.x]: also update docs for the `owner` field of `DataEntry`

File: docs/paths/update_data_owner_op.yaml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+      - Data
+    summary: Update an owner of data
+    description: |
+                  Allows to update an owner of data
+
+                  #### Account Role Requirements
+                    | Account | Resource Type | Rule Action | Condition | Details |
+                    | --------|:-------------:|:-----------:| :----: |:-----: |
+                    | source  | `DATA`        | `UPDATE_OWNER`    | source `accountID` is equal to data owner | Resource contains data type |
+                    | source  | `DATA`        | `UPDATE_OWNER_FOR_OTHER`    | source `accountID` is **not** equal to data owner | Resource contains data type |
```
Comment (1henny) [inline docs/paths/update_data_owner_op.yaml]: - rule action should be named `TRANSFER_OWNERSHIP`
- we should not have `*_FOR_OTHER` action for this operation as it opens the probable attack vector
- add `TRANSFER_OWNERSHIP` to resource-account-rule.x and resource-signer-rule.x

File: operation-update-data-owner.x [hunk 4]
```diff
@@ -0,0 +1,37 @@
+%#include "xdr/types.h"
+
+namespace stellar
+{
+
+struct UpdateDataOwnerOp
```
Comment (1henny) [inline operation-update-data-owner.x:6]: we also need a reviewable request for this operation

File: operation-update-data-owner.x [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    //: Reserved for future extension
+    EmptyExt ext;
+};
+
+enum UpdateDataOwnerResultCode
+{
+    //: An owner of the data was successfully updated
+    SUCCESS = 0,
+    //: `owner` must be in a valid format
+    INVALID_DATA = -1,
```
Comment (1henny) [inline operation-update-data-owner.x]: `AccountID` type restricts the range of values that may be used in this field so this error code is not needed

**Review by 1henny — CHANGES_REQUESTED** [review]

File: ledger-entries-reviewable-request.x [hunk 6]
```diff
@@ -133,6 +134,8 @@ struct ReviewableRequestEntry {
             CreateDeferredPaymentRequest createDeferredPaymentRequest;
         case CLOSE_DEFERRED_PAYMENT:
             CloseDeferredPaymentRequest closeDeferredPaymentRequest;
+		case DATA_OWNER_UPDATE:
```
Comment (1henny) [inline ledger-entries-reviewable-request.x]: fix indentation

File: operation-create-data-owner-update-request.x [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+    //: Reserved for future extension
+    EmptyExt ext;
+};
+
+enum CreateDataOwnerUpdateRequestResultCode
+{
+    SUCCESS = 0,
+    INVALID_VALUE = -1,
+    UPDATE_DATA_TASKS_NOT_FOUND = -2,
```
Comment (1henny) [inline operation-create-data-owner-update-request.x]: update_data_tasks are related to update_data operation and not to update_data_owner. I think it should not be here

**Review by 1henny — CHANGES_REQUESTED** [review]

File: resource-account-rule.x [hunk 8]
```diff
@@ -153,6 +153,14 @@ case DATA_UPDATE:
         //: Reserved for future extension
         EmptyExt ext;
     } dataUpdate;
+case DATA_OWNER_UPDATE:
+    struct
+    {
+        //: Numeric type of the data
+        uint64 type;
```
Comment (artemskriabin) [inline resource-account-rule.x:160]: it could rule for the reviewable request

File: operation-update-data-owner.x [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    NOT_FOUND = -1,
+    //: Only owner can update data entry
+    NOT_AUTHORIZED = -2
+};
+
+//: Result of successful application of `UpdateDataOwner` operation
+struct UpdateDataOwnerSuccess
+{
+    //: A new owner of the entry
+    AccountID new_owner;
```
Comment (artemskriabin) [inline operation-update-data-owner.x]: new owner is specified in the request

File: reviewable-request-update-data-owner.x [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+{
+struct DataOwnerUpdateRequest {
+    //: Id of the data entry
+    uint64 id;
+
+    //: Sequence number increases when request is rejected
+	uint32 sequenceNumber;
+
+    //: A new owner of data
+    AccountID new_owner;
```
Comment (artemskriabin) [inline reviewable-request-update-data-owner.x]: snake case

**Review by artemskriabin — CHANGES_REQUESTED** [review]

File: ledger-entries-reviewable-request.x [hunk 11]
```diff
@@ -54,7 +54,8 @@ enum ReviewableRequestType
 	DATA_UPDATE = 23,
 	DATA_REMOVE = 24,
 	CREATE_DEFERRED_PAYMENT = 25,
-    CLOSE_DEFERRED_PAYMENT = 26
+    CLOSE_DEFERRED_PAYMENT = 26,
```
Comment (1henny) [inline ledger-entries-reviewable-request.x:58]: fix indentation

File: reviewable-request-update-data-owner.x [hunk 12]
```diff
@@ -0,0 +1,26 @@
+%#include "xdr/ledger-entries-data.h"
+
+namespace stellar
+{
+struct DataOwnerUpdateRequest {
+    //: Id of the data entry
+    uint64 id;
```
Comment (1henny) [inline reviewable-request-update-data-owner.x]: we can use operation field here as in reviewable-request-payment.x

**Review by 1henny — CHANGES_REQUESTED** [review]

File: reviewable-request-update-data-owner.x [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+namespace stellar
+{
+struct DataOwnerUpdateRequest {
+    UpdateDataOwnerOp updateDataOwnerOp;
+
+    //: Sequence number increases when request is rejected
+    uint32 sequenceNumber;
+
+    //: A new owner of data
+    AccountID newOwner;
```
Comment (1henny) [inline reviewable-request-update-data-owner.x]: it is already inside `updateDataOwnerOp`

**Review by 1henny — CHANGES_REQUESTED** [review]

**Review by 1henny — APPROVED** [review]

**Review by artemskriabin — APPROVED** [review]

**Pull request merged**